CMakeLists.txt

```diff
@@ -55,16 +61,20 @@ set(PROCESSOR_IS_AARCH64 FALSE)
 set(PROCESSOR_IS_X86 FALSE)
 set(PROCESSOR_IS_POWER FALSE)
 
-if(CMAKE_SYSTEM_PROCESSOR MATCHES "^mips")
-  set(PROCESSOR_IS_MIPS TRUE)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^arm")
-  set(PROCESSOR_IS_ARM TRUE)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^aarch64")
+if(HOST_ARCHITECTURE MATCHES "^arm64")
   set(PROCESSOR_IS_AARCH64 TRUE)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(AMD64|amd64)|(^i.86$)")
-  set(PROCESSOR_IS_X86 TRUE)
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)")
-  set(PROCESSOR_IS_POWER TRUE)
+else()
+  if(CMAKE_SYSTEM_PROCESSOR MATCHES "^mips")
+    set(PROCESSOR_IS_MIPS TRUE)
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^arm")
+    set(PROCESSOR_IS_ARM TRUE)
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^aarch64")
+    set(PROCESSOR_IS_AARCH64 TRUE)
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(AMD64|amd64)|(^i.86$)")
+    set(PROCESSOR_IS_X86 TRUE)
+  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)")
+    set(PROCESSOR_IS_POWER TRUE)
+  endif()
 endif()
 
 macro(add_cpu_features_headers_and_sources HDRS_LIST_NAME SRCS_LIST_NAME)
```

Comment:
I've been editing things to get this building on M1 (actually Universal) and I had to change the way files are included because generating a project as it stands currently, it'll pick one of the architecture's sets of files to include, but then if you build as universal it obviously does not have the required files for the other architecture. So I changed it locally so that all the files are always included and then added #if per platform in the various platform files.

Reply:
Oh, and there are also issues with including the libraries because it decides based on the **host** architecture for some of it (unix_based_hardware_detection) so it can fail to build in those cases too.

Reply:
Yeah CMake universal builds are a bit of a pain. They are 1 build, making universal targets, so anything done at configure time is probably incorrect. I found that you basically have to move all this stuff ot compile time with ifdefs.

Reply:
my 2 cents,
`CMAKE_SYSTEM_PROCESSOR` is for **target** system and it usually part of a cmake cross-toolchain otherwise you have the `CMAKE_HOST_SYSTEM_PROCESSOR`...
ref: https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_PROCESSOR.html

For universal platform, i don't know if we could/should introduce a "universal" cmake system processor and propagate it accordingly than including everything and using preprocessor everywhere...

Reply:
The way to build for universal currently means invoking xcodebuild with dual architecture. This is well after cmake has decided which files to include. I know the CMake guys have been doing some work on all this, but its still early days. Think we just have to suck it up for now.

Reply:
> The way to build for universal currently means invoking xcodebuild with dual architecture. This is well after cmake has decided which files to include.

I means, instead of adding **all** files we could simply add files supported by their "universal" build, I'm not sure Apple support (mips, ppc, arm, arm64 and x86)...

> I know the CMake guys have been doing some work on all this, but its still early days. Think we just have to suck it up for now.

I think you must talk about [`CMAKE_OSX_ARCHITECTURES`](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html)

Reply:
Yeah, I've tried using that to target a specific target and that's ok, but it does not quite work yet. ie. you can't tell it x86_64;arm64 as it complains. Hence the xcodebuild route.